pkg/apis/networking/types.go

```diff
@@ -235,12 +235,25 @@ type IngressList struct {
 
 // IngressSpec describes the Ingress the user wishes to exist.
 type IngressSpec struct {
-	// A default backend capable of servicing requests that don't match any
-	// rule. At least one of 'backend' or 'rules' must be specified. This field
-	// is optional to allow the loadbalancer controller or defaulting logic to
-	// specify a global default.
+	// Class is the name of the IngressClass cluster resource. The associated
+	// IngressClass defines which controller will implement the resource.
+	// This replaces the deprecated `kubernetes.io/ingress.class` annotation.
+	// For backwards compatibility, when that annotation is set, it must be
+	// given precedence over this field. The controller may emit a warning if
+	// the field and annotation have different values. Implementations of this
+	// API should ignore Ingresses without a class specified.
+	// An IngressClass resource may be marked as default, which can be used to
+	// set a default value for this field. For more information, refer to the
+	// IngressClass documentation.
 	// +optional
-	Backend *IngressBackend
+	Class *string
```

Comment:
Do we also expect ingress controllers to just watch for Ingress objects that match their class name? If so, we should add fieldSelector support for this (so a controller could list/watch `ingresses?fieldSelector=spec.ingressClassName=foo`, for example).

To do that:
1. Make methods in the ingress strategy to:
  1. Build a map of selectable fields for an ingress object ([example](https://github.com/kubernetes/kubernetes/blob/96dfa3f6051766d7ebb3a74dbc0c0c58d5aa120b/pkg/registry/batch/job/strategy.go#L201-L208))
  2. Return that map along with labels for matching field/label selectors ([example](https://github.com/kubernetes/kubernetes/blob/96dfa3f6051766d7ebb3a74dbc0c0c58d5aa120b/pkg/registry/batch/job/strategy.go#L210-L217))
  3. Construct a match predicate from those label/field sets ([example](https://github.com/kubernetes/kubernetes/blob/96dfa3f6051766d7ebb3a74dbc0c0c58d5aa120b/pkg/registry/batch/job/strategy.go#L219-L228))
2. Assign the match predicate function as the ingress storage PredicateFunc ([example](https://github.com/kubernetes/kubernetes/blob/96dfa3f6051766d7ebb3a74dbc0c0c58d5aa120b/pkg/registry/batch/job/storage/storage.go#L63))
3. Call AddFieldLabelConversionFunc for each external version ([example](https://github.com/kubernetes/kubernetes/blob/96dfa3f6051766d7ebb3a74dbc0c0c58d5aa120b/pkg/apis/batch/v1/conversion.go#L30-L41)), mapping the fieldSelector field names to the internal field names. In this case, the external and shared field names would be identical: `metadata.name`, `metadata.namespace`, and `spec.ingressClassName` (or whatever we end up naming the field)

When that's done, the following should work:
```
kubectl get ingresses.v1beta1.extensions --field-selector=metadata.name=$myname
kubectl get ingresses.v1beta1.extensions --field-selector=metadata.namespace=$mynamespace
kubectl get ingresses.v1beta1.extensions --field-selector=spec.ingressClassName=$myclassname

kubectl get ingresses.v1beta1.networking.k8s.io --field-selector=metadata.name=$myname
kubectl get ingresses.v1beta1.networking.k8s.io --field-selector=metadata.namespace=$mynamespace
kubectl get ingresses.v1beta1.networking.k8s.io --field-selector=spec.ingressClassName=$myclassname

kubectl get ingresses.v1.networking.k8s.io --field-selector=metadata.name=$myname
kubectl get ingresses.v1.networking.k8s.io --field-selector=metadata.namespace=$mynamespace
kubectl get ingresses.v1.networking.k8s.io --field-selector=spec.ingressClassName=$myclassname
```

Reply:
Agree that `ingressClassName` works better here, will work on changing that. 

The relationship between class name and controllers is not as straightforward as it might seem. The class name here refers to the name of the `IngressClass` resource, which then specifies the controller name. The idea was to allow for a single controller to implement multiple classes of ingress, such as external and internal load balancers. So in this case, we'd likely want IngressClass to be selectable in this same way. At that point, controllers would watch IngressClass resources with their controller name, then modify their watch on Ingress resources to include any with those class names. Does that sound reasonable?